Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,493 @@
+  ⟨Array.pop v.toArray, by {simp only [size_pop, v.size_eq]}⟩
+
+/--
+`shrink v m` removess the last `m` elements of a vector.
+-/
+def shrink (v : Vector α n) (m : Nat) : Vector α (n - m) :=
+  match m with
+  | 0 => v
+  | (x + 1) =>
+      have h : n - 1 - x = n - (x + 1) := by
+        omega
+      h ▸ (shrink (Vector.pop v) x)
+
+/--
+`isEqvAux a b hsz p i` is a helper function for `isEqv`. It performs
+an index by index check on whether the corresponding elements of a vector
+satisfy the boolean property `p` starting at a given index `i`
+-/
+def isEqvAux (a b : Vector α n)
+  (p : α → α → Bool)
+  (i : Nat) : Bool :=
+  if h : i < size a then
+     have : i < size b := h
+     p a[i] b[i] && isEqvAux a b p (i+1)
+  else
+    true
+termination_by size a - i
+decreasing_by simp_wf; decreasing_trivial_pre_omega
+
+/--
+`isEqv` takes a given boolean property `p`. It returns `true`
+if and only if `p a[i] b[i]` holds true for all valid indices `i`.
+-/
+def isEqv (a b : Vector α n) (p : α → α → Bool) : Bool := isEqvAux a b p 0
+
+instance [BEq α] : BEq (Vector α n) :=
+  ⟨fun a b => isEqv a b BEq.beq⟩
+
+
+/--`reverse v` reverses the vector `v`-/
+def reverse (v : Vector α n) : Vector α n :=
+  ⟨Array.reverse v.toArray, proof⟩
+  where
+    proof := by
+      simp
+      exact v.size_eq
+      done
+/--
+Remove the element at a given index from a vector without bounds checks, using a Fin index.
+
+This function takes worst case O(n) time because it has to backshift all elements at positions
+greater than i.
+-/
+def feraseIdx (v : Vector α n) (i : Fin n) : Vector α (n-1) :=
+  ⟨Array.feraseIdx v.toArray (v.size_eq.symm ▸ i), proof⟩
+  where
+    proof := by
+      simp [Array.size_feraseIdx]
+      rw[v.size_eq]
+
+
+/--
+  `Vector.tail` produces the tail of a vector
+-/
+def tail (v : Vector α n)  : Vector α (n-1):=
+  match n with
+  | 0 => v
+  | _ + 1 => Vector.feraseIdx v 0
+
+
+/-- Remove the element at position `i` from a vector of length `n` if `i < n`. Panics otherwise.
+
+This function takes worst case O(n) time because it has to backshift all elements at positions
+greater than i.
+ -/
+def eraseIdx! (v : Vector α n) (i : Nat) : Vector α (n-1) :=
+    if h : i < n then
+      feraseIdx v ⟨i,h⟩
+    else
+      have _ : Inhabited (Vector α (n-1)) := ⟨v.tail⟩
+      panic! "Index" ++ toString i ++ "is out of bounds"
+
+@[deprecated (since := "2024-05-04")] alias removeNth := eraseIdx
+
+
+/--
+If `x` is an element of vector `v` at index `j` such that `j > i`,
+then `indexOfAux v x i` returns `some j`. Else it returns `none`.
+-/
+def indexOfAux [BEq α] (v : Vector α n) (x : α) (i : Nat) : Option (Fin n) :=
+  match Array.indexOfAux v.toArray x i with
+  | some x => some <| v.size_eq ▸ x
+  | none => none
+
+/--
+If `x` is an element of vector `v` at index `j`,
+then `indexOf? v x` returns `some j`. Else it returns `none`.
+-/
+def indexOf? [BEq α] (v : Vector α n) (x : α) : Option (Fin n) :=
+  indexOfAux v x 0
```

Comment:
```suggestion
def indexOf? [BEq α] (v : Vector α n) (x : α) : Option (Fin n) :=
  match n, v with | _, ⟨a, rfl⟩ => a.indexOf? x
```
The same trick should work elsewhere to get rid of all the `*Aux` definitions.

Reply:
Although these functions are called auxiliary, I think all of them are useful in their own right. Many are more general than the functions they are used to define.

Reply:
Calling the array code is better than copying it. That way the vector code is automatically up to date with array.

Reply:
Then the Vector API isn't complete. Every time someone wants to use an auxiliary function, they also have to deal with Array code

Reply:
My ideal for Vectors is that I shouldn't have to peek underneath to use them as much as possible.

Reply:
All of these `*Aux` functions are poised to disappear in a forthcoming array cleanup in core, so they really don't belong in the vector API. If you believe these functions are useful, first add them to the array API with proper names and then add them to vector.

I general, it's a bad idea to copy the implementation of an array function instead of calling the array function. Remember that Array has deep compiler support, so the implementations are often dummy code just to please the kernel. Also, when the implementation in Array gets an update, the Vector implementation won't update itself and Vector will lag behind. In Batteries, performance and maintainability are always primary concerns.

Reply:
Done.

I removed the aux functions and used Array functions directly. That being said, I think there were many private functions like isPrefixOfAux, isEqvAux, and the diff Aux functions which I will find myself writing anyway. It might make sense salvage them in a later PR for both Arrays and Vectors. The problem is I don't want to touch core with its convoluted PR process.